packages/utils/src/object.js

```diff
@@ -192,3 +194,13 @@ export const filterValues = (object, valueFilter) =>
 
 export const stripFields = (object = {}, fieldsToExclude = []) =>
   Object.fromEntries(Object.entries(object).filter(([key]) => !fieldsToExclude.includes(key)));
+
+const sortFields = object => {
+  const sortedEntries = Object.entries(object).sort(([keyA], [keyB]) => compareAsc(keyA, keyB));
+  return Object.fromEntries(sortedEntries);
```

Comment:
I don't understand this but maybe you could explain it with an example? Or in a call

Reply:
@mcccclean can you confirm whether I'm right here? 

Starting with two objects with the same key/value pairs (none of which are nested objects) but in undefined orders
- get the entries out of the object
- sort them by key
- construct a new object from those sorted entries
- JSON.stringify the object
If you then compare the results, are they guaranteed to match? I am thinking no, because when you construct the new object from the sorted entries, the order of the entries in the new object is not guaranteed?

Reply:
@edmofro perhaps you refer to [Object.entries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries)? The docs for it mention

> The order of the array returned by Object.entries() does not depend on how an object is defined

---

For `Object.fromEntries`:

According to its [specification](https://tc39.es/ecma262/#sec-object.fromentries), it uses the [AddEntriesFromIterable](https://tc39.es/ecma262/#sec-add-entries-from-iterable) algorithm to create the new object. My understanding is that this will result in properties to be inserted in the new object in the same order they are defined in the entries array.

On the other hand, `JSON.stringify` doesn't guarantee property order (see [this note](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#issue_with_plain_json.stringify_for_use_as_javascript)).  However, I would assume that for an object with the same props defined in the same order we'll always get the same string, even if the fields are re-ordered in the stringified result. (This is not relevant directly to this function, but it is to [getUniqueObjects](https://github.com/beyondessential/tupaia/blob/1779-flatten-data-fetching/packages/utils/src/object.js#L203) which uses it below)

Reply:
My understanding was just that in general, objects never guarantee order of properties

Reply:
However if you're using the same exact operations to construct and then stringify the object, I agree that you would expect that to be deterministic, and the result should be the same (even if the order isn't the same as the order they were sorted into)

Reply:
Yes, that would work for `getUniqueObjects`. However, `sortFields` could still be a misnomer for this function, if the enumeration order does not match our intended field sort order.

I believe in this specific case this happens to be fine though. We use `compareAsc` for sorting, which will place all integers sorted by value first (1 before 10 etc), then all alphanumerics sorted in ASC order. This matches the enumeration order introduced in ES6 according to [this article](https://dev.to/eladtzemach/traversal-order-of-object-properties-in-javascript-es6-5a3) (haven't found an official doc yet!)

So while in the general case the insertion order is not preserved during enumeration, in this case our sorting logic just happens to be compatible with the prop enumeration logic, which is a happy coincidence 😆

Reply:
Weighing in - my understanding is that property order in an object [_is_ predictable](https://www.stefanjudis.com/today-i-learned/property-order-is-predictable-in-javascript-objects-since-es2015/), but only since ES2015, and that that assumption might not survive transpilation to an earlier spec? So if this is getting transpiled to support older browsers it might break (and it would break on newer browsers too, because it'd be running the same transpiled code as an old IE or whatever).

BUT, the transpilation process is itself predictable! So, like, given that this particular code is running correctly now, I would feel pretty safe to assume that it runs correctly in general.

Reply:
Cool cool, it sounds like we all agree that this is totally fine in practice :-)

Reply:
That was an interesting discussion! It made me realise that we don't handle string and number comparison correctly in `compareAsc`. We used `<` and `>`, which always return `false` when a string and a number are compared, regardless of the operand order.

This is now fixed in the latest [commit](https://github.com/beyondessential/tupaia/pull/1873/commits/6148c3744ff7cac4dc832b67f27fa3d62a707bee). I also added a compatibility note in `sortFields`